### ISSUE TITLE
NativeLog causes crash when changing MRL

### ISSUE THREAD

**Issue by lukasHoel** [caprica/vlcj#818]
Hi,
I am currently migrating to VLCJ 4.0.0 and have noticed a strange behaviour with `NativeLog`.
The following small sample program causes a deadlock / crash in video playback. Sometimes, the JVM even crashed (not reproducable through).

```
EmbeddedMediaPlayerComponent embeddedMediaPlayerComponent = new EmbeddedMediaPlayerComponent();
EmbeddedMediaPlayer embeddedMediaPlayer = embeddedMediaPlayerComponent.mediaPlayer();
		
NativeLog nativeLog = embeddedMediaPlayerComponent.mediaPlayerFactory().application().newLog();
nativeLog.addLogListener(new LogEventListener() {		
	@Override
	public void log(LogLevel level, String module, String file, Integer line, String name, String header, Integer id, String message) {
		System.out.println(message);	
	}
});

JFrame frame = new JFrame();
frame.add(embeddedMediaPlayerComponent);
frame.setSize(800, 480);
frame.setVisible(true);

// User starts playing the first video
embeddedMediaPlayer.media().play("videoA");

// Simulate some wait time - user watches the video
Thread.sleep(5000);

// User changes to another video in the same player manually
embeddedMediaPlayer.media().play("videoB");

// This does not fix the problem
embeddedMediaPlayer.submit(() -> {
	embeddedMediaPlayer.media().play("videoB");
});

```
When I do not create a `NativeLog ` object then this program works!

The following exception gets logged by JNA sometimes when the JVM crash happens:

```
Feb 27, 2019 3:42:01 PM com.sun.jna.Native$1 uncaughtException
WARNUNG: JNA: Callback java.lang.Object@54eeda70 threw the following exception
java.lang.IncompatibleClassChangeError
```

Setup:
VLCJ 4.0.0
VLC 3.0.4
Windows 64 Bit
Java 8 SE

I would appreciate any help! Am I using VLCJ the wrong way or is this an internal problem?

Thanks,
Lukas

**Comment by caprica** [comment]
Well the IncompatibleClassChangeError is very suspicious, are you running in a debugger and recompiling?

For the hard crash - more often than not, this is because your Java object got garbage collected. Are you absolutely certain you are keeping a hard reference to the NativeLog instance?

I would also encourage you to try the native log in the vlcj-player application to see if it also crashes for you.

And when you say a deadlock, what precisely do you mean? Is your UI thread blocked? Or something else?

**Comment by caprica** [comment]
Oh wow, I just tested vlcj-player and indeed it deadlocks when the native log is active and you change media.

**Comment by caprica** [comment]
It will actually deadlock just invoking stop, let alone changing media.

Try setting any log level other than DEBUG, so NOTICE, WARNING or ERROR.

I can consistently get a deadlock when the level is DEBUG, I never saw any problem with the other log levels.

I'm not saying this is the answer, it's just a data-point.

It feels like it might be a native issue.

**Comment by lukasHoel** [comment]
> Well the IncompatibleClassChangeError is very suspicious, are you running in a debugger and recompiling?

I start the provided test program from inside Eclipse. When the deadlock occurs, I have to manually terminate the Java program. Maybe this yields resource leaks which cause that error on subsequent tries, I dont know.

> Try setting any log level other than DEBUG, so NOTICE, WARNING or ERROR.
> I can consistently get a deadlock when the level is DEBUG, I never saw any problem with the other log levels.

Using the test program I provided in my first message, **I still get the same behavior** when changing the LogLevel manually to any of the available values. I attach that program in this comment, maybe you can try it yourself with that program to confirm or deny the issue on your system? Furthermore, what LibVLC version are you using? As I told previously, I am using 3.0.4 (Vetinari, 64-Bit on Windows), which is not the newest available version. When this is indeed a native issue this might also cause different behaviour?


[NativeLog-Crash-Sample-Program.txt](https://github.com/caprica/vlcj/files/2913936/NativeLog-Crash-Sample-Program.txt)



> For the hard crash - more often than not, this is because your Java object got garbage collected. Are you absolutely certain you are keeping a hard reference to the NativeLog instance?

In the attachement, I changed the reference to be a static class variable which gets printed at the end of the program. So this should definitely make sure to not garbage collect that object.

**Comment by caprica** [comment]
Thanks for the report.

I tested VLC 3.0.4 and VLC 3.0.6 on Linux.

I don't doubt you, but it's odd that if I don't use DEBUG level it works for me, but not for you. I tested it *many* times and the behaviour was consistent. 

I'll try your sample as soon as I can.

By the way, did you try this with vlcj 3.x previously?

**Comment by caprica** [comment]
In your example program, I always feel more comfortable about these type of bugs if you put a Thread.currentThread.join() at the end of your main method. Just to 100% eliminate the possibility of garbage-collection-induced crashes.

**Comment by caprica** [comment]
OK, when I run your test, I get a lock-up every single time when the log level is debug.

Changing it to notice, I never get a single lock-up.

It might just be coincidence but so far it is 100%.

Again, I accept this is not a solution, but it's odd for sure.

**Comment by caprica** [comment]
Pretty obvious deadlock here in fact:
```
"log-events" #27 daemon prio=5 os_prio=0 tid=0x00007f2a54007800 nid=0x3428 waiting for monitor entry  [0x00007f2a5989d000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.sun.jna.Native$3.invoke(Native.java:1195)
        - waiting to lock <0x00000005e916f418> (a com.sun.jna.NativeLibrary)
      *  at com.sun.proxy.$Proxy3.libvlc_log_get_context(Unknown Source) *
        at uk.co.caprica.vlcj.log.NativeLog$NativeLogCallback.log(NativeLog.java:179)
        at jdk.internal.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@10.0.2/DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(java.base@10.0.2/Method.java:564)
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.invokeCallback(CallbackReference.java:520)
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.callback(CallbackReference.java:551)
```
```
"log-events" #33 daemon prio=5 os_prio=0 tid=0x00007f2a20006000 nid=0x3432 waiting for monitor entry  [0x00007f2a36acc000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.sun.jna.Native$3.invoke(Native.java:1195)
        - waiting to lock <0x00000005e916f418> (a com.sun.jna.NativeLibrary)
      * at com.sun.proxy.$Proxy3.libvlc_log_get_context(Unknown Source) *
        at uk.co.caprica.vlcj.log.NativeLog$NativeLogCallback.log(NativeLog.java:179)
        at jdk.internal.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@10.0.2/DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(java.base@10.0.2/Method.java:564)
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.invokeCallback(CallbackReference.java:520)
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.callback(CallbackReference.java:551)
```

**Comment by caprica** [comment]
So, the reason this is happening in vlcj-4...

We have two callback threads running concurrently for the native log, I do not know why this is happening - it seems like VLC might be sending us a log message while we are still processing an existing one.

Anyway, both those callbacks are invoking the same method on the LibVlc native library instance.

In vlcj-4, the native library is by default synchronised (wrapped it a JNA synchronisation facade). In vlcj-3 the native library was not synchronised.

All the stuff about DEBUG vs NOTICE was just due to races I suppose.

**Comment by caprica** [comment]
#751

**Comment by caprica** [comment]
Further investigation, it's actually a deadlock between the native methods invoked by the play call and native log context call.

The conclusion is the same, it's because the native library is synchronised.

This was never something I was completely happy with, it was difficult to find conclusive information on whether re-entering LibVLC in this manner was safe. Although for the entire lifetime if vlcj-3, I did NOT use a synchronised instance.

I think I have to revert that synchronised library change.

**Comment by caprica** [comment]
The code change will be tracked under #820.

Thanks again for taking the time to report this and to provide a decent test case.

**Comment by lukasHoel** [comment]
Thanks for your detailed responses! Not using the NativeLog is no dealbreaker for me, so I will just disable it until the next release. Btw this did not happen in vlcj-3 for me.

**Comment by caprica** [comment]
It's definitely because of the native library synchronisation done in vlcj-4, that's why vlcj-3 didn't have the same issue.

I will push a fix to Maven Central soon, look out for a 4.0.2 release today/tomorrow.
